src/core/mempool.rs

```diff
@@ -1064,6 +1173,111 @@ impl MemPoolDB {
         Ok(updated)
     }
 
+    /// Returns an iterator over the mempool entries that do have a fee rate, sorted by fee rate.
+    /// Page size is 10_000. TODO: Make this configurable.
+    fn sorted_fee_rate_transactions(
+        connection: Arc<Mutex<Connection>>,
+    ) -> Box<dyn Iterator<Item = MemPoolTxMinimalInfo>> {
+        info!("sorted_fee_rate_transactions");
+        let sql = "
+        SELECT txid, origin_nonce, origin_address, sponsor_nonce, sponsor_address, fee_rate
+        FROM mempool
+        WHERE fee_rate IS NOT NULL
+        ORDER BY fee_rate DESC
+        LIMIT ?
+        OFFSET ?
```

Comment:
You probably didn't see this yet because you're working with a small database, but `OFFSET` will do a full table scan up to the given offset value.  Multiple calls to this iterator for `n` transactions will have `O(n^2)` runtime complexity.  You should avoid using `OFFSET` altogether, and instead use a `WHERE` clause on an indexed column so that database reads can begin at the right page.

Alternatively, you could issue just this query with no offset, and keep around the Rusqlite query statement for the lifetime of the iterator, and just read rows out as `.next()` is called.  You're going to do `Omega(n)` work anyway because you're doing a full scan of the DB, but we really should avoid `O(n^2)` work with multiple calls to a query with `OFFSET`.

Reply:
These are all good suggestions. I like the second idea.

However, the `SELECT` part in and of itself isn't that expensive I think. It's 1) getting the data into Rust and 2) deserializing transactions (aside from 3) processing) that takes a lot of time.

In other words, this wasted `SELECT` time has a really small constant factor.

At some point, we should just merge something using this iterator paradigm and we can follow up on that with little improvements.. Like we have #3321 that builds on this, we can make and measure one adjustment at a time.

Reply:
I'll try to formulate it the DB cursor way and see if it's faster/simpler.

Reply:
@gregorycoppola The use of `OFFSET` itself is going to make iteration `O(n^2)` because it's slow in sqlite.  If you're reading a 10,000 row table, and you do a `SELECT ... FROM ... WHERE .. OFFSET 9999`, sqlite will read the first 9998 rows' contents and discard them because that's just the way the RDBMS is built.  See [here](https://stackoverflow.com/questions/12266025/sqlite-query-optimization-using-limit-and-offset) for details.

We really shouldn't be using `OFFSET` anywhere in the code if we can help it.

Reply:
But, the overall cost of that scan is neglible, otherwise we wouldn't have good performance numbers on @obycode 's profiling experiments.

I was foiled on every attempt at getting `rusqlite` to create an iterator. It is so hard. Maybe someone else can get it to work but I found it insanely hard.

The `rusqlite` library uses patterns th at create downstream problems, I think, and we should use other models for how to write modern Rust code.

Reply:
There is another pattern I think where you can use an index and a cursor.

The data set is around 28k in the mempoool, from #3313:

Mempool breakdown:

    Total size: 25214
    Nonce less than expected (already processed): 21308
    Nonce is expected (to be processed): 1791
    Nonce greater than expected(can not process yet): 2115

Reply:
Oh yeah, you already mentioned the cursor pattern as the first alternative.

Reply:
How would this work with the cursor pattern?

**problem**
If we sort by fee rate, then, it's possible there are multiple transactions with the same fee rate. E.g., 

```
[{fee_rate 10, id=1}, {fee_rate = 10, id = 2}, {fee_rate = 9, id=3}]
```

If the last tx processed was `id=1`, and you do the `WHERE` clause as `WHERE fee_rate < 10`, then you will miss 2. If you do `WHERE fee_rate <= 10`, then you will hit `id=1` again.

**idea**
combine `fee_rate` and `time_received` into a single field, and sort by:
* higher fee rate will win
* if tied on fee rate older `time_receivd` will win

**implementation**
Encode `fee_rate` and `time_received` into a single value, and sorting by that. We would have to either add a new field, or repurpose `fee_rate` to be this new field.


Or how would you guys do this?

Reply:
> Encode fee_rate and time_received into a single value

No need to do this.  Just have a compound `ORDER BY` and `WHERE` clauses, like `WHERE fee_rate < ? OR (fee_rate == ? AND time_received < ? ORDER BY fee_rate DESC, time_received ASC`) where `fee_rate` and `time_received` are both indexed columns.  Your cursor is then simply the last `fee_rate` and `time_received` you saw.  Whatever you do, be sure to run `EXPLAIN QUERY PLAN` on the query to make _absolutely certain_ that you are not introducing any table scans.

Reply:
@gregorycoppola I have a clean solution to these issues that I was testing over the weekend but have been delayed by a family health issue. I think I should be able to chat later today if you are available and I can share that progress.

Reply:
@obycode ok nice!.. this is the last issue remaining.

Reply:
Hello everyone...

I have been doing some profiling to see how time taken to read in entries actually empirically changes as the `OFFSET` grows.

I pushed the commit that does the logging

We **do not** see evidence that the time taken to read in examples is actually even growing at all as `OFFSET` grows.

```
offset: 0, avg: 20916.010055478502, count: 2884
offset: 10000, avg: 8705.302325581395, count: 2838
offset: 20000, avg: 8698.485935984481, count: 2062
offset: 30000, avg: 15763.23076923077, count: 793
offset: 40000, avg: 17311.821292775665, count: 526
offset: 50000, avg: 16809.42105263158, count: 380
offset: 60000, avg: 13563.802973977696, count: 269
offset: 70000, avg: 7634.550898203593, count: 167
offset: 80000, avg: 2070.4074074074074, count: 81
```

Note how offset `0` has the largest average. Also, the average for `80000` offset is actually smaller than the others.

I think randomness, and IO factors are more significant.

This is important because, if we implement paging, it will:
* complicate the code
* **introduce new costs**

Thus, not only is this `OFFSET`-based solution the simplest, it will probably turn out to be the fastest.